### PR TITLE
21077 Add support for custom ports in multisite site addresses

### PR DESCRIPTION
**Pull request by soderlind**
Fix bug in patch 15936.5.diff, re-add the port number in ms-site.php

Trac ticket: https://core.trac.wordpress.org/ticket/21077

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-admin/includes/network.php [hunk 1]
```diff
@@ -138,7 +138,10 @@ function network_step1( $errors = false ) {
 
 	$hostname  = get_clean_basedomain();
 	$has_ports = strstr( $hostname, ':' );
-	if ( ( false !== $has_ports && ! in_array( $has_ports, array( ':80', ':443' ), true ) ) ) {
+	/** This filter is documented in wp-includes/ms-settings.php */
+	$allowed_ports = apply_filters( 'allowed_multisite_ports', array( ':80', ':443' ) );
+
+	if ( ( false !== $has_ports && ! in_array( $has_ports, $allowed_ports ) ) ) {
```
Comment (costdev) [inline src/wp-admin/includes/network.php]:
```suggestion
	if ( ( false !== $has_ports && is_array( $allowed_ports ) && ! in_array( $has_ports, $allowed_ports, true ) ) ) {
```
- We call `in_array()` without confirming that `$allowed_ports` is an array after filtering. We should add this guard.
- Adding `true` to `in_array()` will make the check strict. `in_array()` should perform a strict check where possible.

File: src/wp-admin/includes/network.php [hunk 2]
```diff
@@ -179,7 +182,7 @@ function network_step1( $errors = false ) {
 	}
 	?>
 	<p><?php _e( 'Welcome to the Network installation process!' ); ?></p>
-	<p><?php _e( 'Fill in the information below and you&#8217;ll be on your way to creating a network of WordPress sites. Configuration files will be created in the next step.' ); ?></p>
+	<p><?php _e( 'Fill in the information below and you&#8217;ll be on your way to creating a network of WordPress sites. We will create configuration files in the next step.' ); ?></p>
```
Comment (costdev) [inline src/wp-admin/includes/network.php]: This change is not related to this ticket. 

Changes were made in the codebase and it seems the initial patch for this ticket was not refreshed on trunk. This change should be removed from this PR.

File: src/wp-admin/includes/network.php [hunk 3]
```diff
@@ -271,7 +274,7 @@ function network_step1( $errors = false ) {
 		<?php
 		printf(
 			/* translators: 1: Site URL, 2: Host name, 3: www. */
-			__( 'You should consider changing your site domain to %1$s before enabling the network feature. It will still be possible to visit your site using the %3$s prefix with an address like %2$s but any links will not have the %3$s prefix.' ),
+			__( 'We recommend you change your site domain to %1$s before enabling the network feature. It will still be possible to visit your site using the %3$s prefix with an address like %2$s but any links will not have the %3$s prefix.' ),
```
Comment (costdev) [inline src/wp-admin/includes/network.php]: This change is not related to this ticket. 

Changes were made in the codebase and it seems the initial patch for this ticket was not refreshed on trunk. This change should be removed from this PR.

File: src/wp-admin/includes/network.php [hunk 4]
```diff
@@ -444,23 +447,23 @@ function network_step2( $errors = false ) {
 			echo '<strong>' . __( 'Caution:' ) . '</strong> ';
 			printf(
 				/* translators: 1: wp-config.php, 2: .htaccess */
-				__( 'You should back up your existing %1$s and %2$s files.' ),
+				__( 'We recommend you back up your existing %1$s and %2$s files.' ),
```
Comment (costdev) [inline src/wp-admin/includes/network.php]: This change is not related to this ticket. 

Changes were made in the codebase and it seems the initial patch for this ticket was not refreshed on trunk. This change should be removed from this PR.

File: src/wp-admin/includes/network.php [hunk 5]
```diff
@@ -444,23 +447,23 @@ function network_step2( $errors = false ) {
...
 				'<code>wp-config.php</code>',
 				'<code>.htaccess</code>'
 			);
 		} elseif ( file_exists( $home_path . 'web.config' ) ) {
 			echo '<strong>' . __( 'Caution:' ) . '</strong> ';
 			printf(
 				/* translators: 1: wp-config.php, 2: web.config */
-				__( 'You should back up your existing %1$s and %2$s files.' ),
+				__( 'We recommend you back up your existing %1$s and %2$s files.' ),
```
Comment (costdev) [inline src/wp-admin/includes/network.php]: This change is not related to this ticket. 

Changes were made in the codebase and it seems the initial patch for this ticket was not refreshed on trunk. This change should be removed from this PR.

File: src/wp-admin/includes/network.php [hunk 6]
```diff
@@ -444,23 +447,23 @@ function network_step2( $errors = false ) {
...
 				'<code>wp-config.php</code>',
 				'<code>web.config</code>'
 			);
 		} else {
 			echo '<strong>' . __( 'Caution:' ) . '</strong> ';
 			printf(
 				/* translators: %s: wp-config.php */
-				__( 'You should back up your existing %s file.' ),
+				__( 'We recommend you back up your existing %s file.' ),
```
Comment (costdev) [inline src/wp-admin/includes/network.php]: This change is not related to this ticket. 

Changes were made in the codebase and it seems the initial patch for this ticket was not refreshed on trunk. This change should be removed from this PR.

File: src/wp-includes/ms-site.php [hunk 7]
```diff
@@ -977,7 +986,7 @@ function clean_blog_cache( $blog ) {
 	 *
 	 * @since 4.6.0
 	 *
-	 * @param string  $id              Site ID as a numeric string.
+	 * @param int     $id              Blog ID.
```
Comment (costdev) [inline src/wp-includes/ms-site.php]: This change is not related to this ticket. 

Changes were made in the codebase and it seems the initial patch for this ticket was not refreshed on trunk. This change should be removed from this PR.

File: src/wp-includes/ms-settings.php [hunk 8]
```diff
@@ -54,15 +54,23 @@
...
-	} elseif ( ':443' === substr( $domain, -4 ) ) {
-		$domain               = substr( $domain, 0, -4 );
-		$_SERVER['HTTP_HOST'] = substr( $_SERVER['HTTP_HOST'], 0, -4 );
+	/**
+	 * Filters allowed HTTP ports in Multisite.
+	 *
+	 * @param array $allowed_ports The allowed ports. Default is [ ':80', ':443' ].
+	 *
+	 * @since 5.3.0
+	 *
```
Comment (costdev) [inline src/wp-includes/ms-settings.php]:
```suggestion
	 * @since 5.3.0
	 *
	 * @param array $allowed_ports The allowed ports. Default is [ ':80', ':443' ].
```

- `@param` then `@since`.
- Removed extra DocBlock line.

Note: If the ticket is milestoned, this `@since` annotation will need to be updated to match the Milestone.

File: src/wp-includes/ms-site.php [hunk 9]
```diff
@@ -498,7 +498,16 @@ function wp_normalize_site_data( $data ) {
...
 	if ( array_key_exists( 'domain', $data ) ) {
 		$data['domain'] = trim( $data['domain'] );
+		$has_ports = strstr( $data['domain'], ':' );
+
+		$data['domain'] = preg_replace( '|:\d+$|', '', $data['domain'] ); // Remove ports.
 		$data['domain'] = preg_replace( '/\s+/', '', sanitize_user( $data['domain'], true ) );
+		/** This filter is documented in wp-includes/ms-settings.php */
+		$allowed_ports = apply_filters( 'allowed_multisite_ports', array( ':80', ':443' ) );
+		if ( ( false !== $has_ports && in_array( $has_ports, $allowed_ports ) ) ) {
+			$data['domain'] = $data['domain'] . $has_ports;
```
Comment (costdev) [inline src/wp-includes/ms-site.php]:
```suggestion
			$data['domain'] .= $has_ports;
```
We can use the `.=` operator here instead of repeating `$data['domain']`.

File: src/wp-includes/ms-site.php [hunk 10]
```diff
@@ -498,7 +498,16 @@ function wp_normalize_site_data( $data ) {
 	// Sanitize domain if passed.
 	if ( array_key_exists( 'domain', $data ) ) {
 		$data['domain'] = trim( $data['domain'] );
+		$has_ports = strstr( $data['domain'], ':' );
+
+		$data['domain'] = preg_replace( '|:\d+$|', '', $data['domain'] ); // Remove ports.
```
Comment (costdev) [inline src/wp-includes/ms-site.php]:
```suggestion
		// Remove ports.
		$data['domain'] = preg_replace( '|:\d+$|', '', $data['domain'] );
```

Smaller screens require horizontal scrolling to see this comment. Let's move it above the line to make things a little easier for those browsing the code.

File: src/wp-includes/ms-settings.php [hunk 11]
```diff
@@ -54,15 +54,23 @@
...
+	 */
+	$allowed_ports = apply_filters( 'allowed_multisite_ports', array( ':80', ':443' ) );
+
+	foreach ( $allowed_ports as $allowed_port ) {
+		$str_length = strlen( $allowed_port );
+		if ( substr( $domain, - $str_length ) == $allowed_port ) {
```
Comment (costdev) [inline src/wp-includes/ms-settings.php]:
```suggestion
		if ( substr( $domain, - $str_length ) === $allowed_port ) {
```

Does this need to be a loose check? If not, then we should use `===` here

File: src/wp-includes/ms-site.php [hunk 12]
```diff
@@ -498,7 +498,16 @@ function wp_normalize_site_data( $data ) {
...
 		$data['domain'] = preg_replace( '/\s+/', '', sanitize_user( $data['domain'], true ) );
+		/** This filter is documented in wp-includes/ms-settings.php */
+		$allowed_ports = apply_filters( 'allowed_multisite_ports', array( ':80', ':443' ) );
+		if ( ( false !== $has_ports && in_array( $has_ports, $allowed_ports ) ) ) {
```
Comment (costdev) [inline src/wp-includes/ms-site.php]:
```suggestion
		if ( ( false !== $has_ports && is_array( $allowed_ports ) && in_array( $has_ports, $allowed_ports, true ) ) ) {
```
- We call `in_array()` without confirming that `$allowed_ports` is an array after filtering. We should add this guard.
- Adding `true` to `in_array()` will make the check strict. `in_array()` should perform a strict check where possible.

File: src/wp-includes/ms-settings.php [hunk 13]
```diff
@@ -54,15 +54,23 @@
...
+	 * Filters allowed HTTP ports in Multisite.
+	 *
+	 * @param array $allowed_ports The allowed ports. Default is [ ':80', ':443' ].
+	 *
+	 * @since 5.3.0
+	 *
+	 */
+	$allowed_ports = apply_filters( 'allowed_multisite_ports', array( ':80', ':443' ) );
+
+	foreach ( $allowed_ports as $allowed_port ) {
```
Comment (costdev) [inline src/wp-includes/ms-settings.php]: We don't know that `$allowed_ports` is an array after filtering.

We should add a guard around the `foreach()`:

```php

if ( is_array( $allowed_ports ) ) {
...
}

```

**Review by costdev — CHANGES_REQUESTED** [review]
@soderlind Thanks for the PR and I noted that this is your first patch to WordPress Core - Welcome!

This review:

- highlights some changes that are not related to this ticket and should be removed from this PR;
- suggests one change for readability;
- suggests one change for documentation;
- suggests changes for stability.

I also note that [patch 15936.5.diff](https://core.trac.wordpress.org/attachment/ticket/21077/15936.5.diff), on which this PR is based, included PHPUnit tests for `tests/phpunit/tests/multisite/site.php`. These tests should also be added to this PR.

File: src/wp-includes/ms-settings.php (hunk 13 above)
Comment (soderlind) [reply src/wp-includes/ms-settings.php]: @costdev Instead of wrapping it, is it, per WordPress standards, ok if I cast the array?
```php
foreach ( (array) $allowed_ports as $allowed_port ) {
```

**Comment by soderlind** [comment]
@costdev I'm looking into why the unit tests failed.

File: src/wp-includes/ms-settings.php [hunk 14]
```diff
@@ -54,15 +54,24 @@
...
+	 * @param array $allowed_ports The allowed ports. Default is [ ':80', ':443' ].
+	 *
+	 * @since 5.3.0
+	 */
+	$allowed_ports = apply_filters( 'allowed_multisite_ports', array( ':80', ':443' ) );
+
+	if ( is_array( $allowed_ports ) ) {
+		foreach ( $allowed_ports as $allowed_port ) {
+			$str_length = strlen( $allowed_port );
+			if ( substr( $domain, - $str_length ) === $allowed_port ) {
```
Comment (costdev) [inline src/wp-includes/ms-settings.php]:
```suggestion
			if ( substr( $domain, - $str_length ) === $allowed_port ) {
```
As of WordPress 5.9, we have a polyfill for `str_ends_with()`. We can use this to improve readability.

```suggestion
			if ( str_ends_with( $domain, $allowed_port ) ) {
```

**Comment by soderlind** [comment]
@costdev New unit test added.

File: src/wp-includes/ms-site.php [hunk 15]
```diff
@@ -977,7 +987,7 @@ function clean_blog_cache( $blog ) {
 	 *
 	 * @since 4.6.0
 	 *
-	 * @param string  $id              Site ID as a numeric string.
+	 * @param string  $id Site ID as a numeric string.
```
Comment (SergeyBiryukov) [inline src/wp-includes/ms-site.php]: This change is not related to this ticket, and appears to not follow the documentation standards.

File: src/wp-includes/ms-settings.php [hunk 16]
```diff
@@ -54,15 +54,24 @@
...
-		$_SERVER['HTTP_HOST'] = substr( $_SERVER['HTTP_HOST'], 0, -3 );
-	} elseif ( ':443' === substr( $domain, -4 ) ) {
-		$domain               = substr( $domain, 0, -4 );
-		$_SERVER['HTTP_HOST'] = substr( $_SERVER['HTTP_HOST'], 0, -4 );
+	/**
+	 * Filters allowed HTTP ports in Multisite.
+	 *
+	 * @param array $allowed_ports The allowed ports. Default is [ ':80', ':443' ].
+	 *
+	 * @since 5.3.0
```
Comment (SergeyBiryukov) [inline src/wp-includes/ms-settings.php]: The `@since` tag should come before `@param`, and would need to be updated to 6.0.0.

File: tests/phpunit/tests/multisite/site.php [hunk 17]
```diff
@@ -1674,6 +1680,18 @@ public function data_wp_normalize_site_data() {
...
+						'domain' => 'domainwithport.com:443',
+						'domain' => 'anotherdomainwithport.com:80',
+						'domain' => 'anotherwithport.com:8080',
+					),
+					array(
+						'domain' => 'domainwithport.com:443',
+						'domain' => 'anotherdomainwithport.com:80',
+						'domain' => 'anotherwithport.com:8080',
+					),
+				),
```
Comment (costdev) [inline tests/phpunit/tests/multisite/site.php]: The `'domain'` key will be overwritten with the last value in each of these arrays.

These should be separated, see the `<yet>/another-invalid-domain.com` test above for an example.

```suggestion
				array(
					array(
						'domain' => 'domainwithport.com:443',
					),
					array(
						'domain' => 'domainwithport.com:443',
					),
				),						
				array(
					array(
						'domain' => 'anotherdomainwithport.com:80',
					),
					array(
						'domain' => 'anotherdomainwithport.com:80',
					),
				),	  
				array(
					array(
						'domain' => 'anotherwithport.com:8080',
					),
					array(
						'domain' => 'anotherwithport.com:8080',
					),
				),
```

**Review by costdev — CHANGES_REQUESTED** [review]

**Review by costdev — APPROVED** [review]
LGTM 👍
